### PR TITLE
Convert DSTP to Python and make it part of the pipeline

### PR DESCRIPTION
**Pull request by bryan-skylight**
# Description
This PR takes the core functionality from the DSTP Kotlin function and converts it into Python code, which could potentially be used in the IntakePipeline setup. It would require some slight modifications, including reading in HL7 instead of FHIR bundles, parsing out the individual messages, and then processing each message instead of each pre-generated FHIR bundle, but the value add is that it moves away from the blob trigger approach that currently isn't working.

# Feedback Requested
* Does this approach seem feasible?  
What are some of the tradeoffs of this approach?  
  * An obvious one is that the IntakePipeline function will take even longer to run, which isn't great given that it already times out  
* Are there other changes that would need to be made other than what's been called out in the code?

# To Do
There's still quite a bit to be done if we were to go down this path (including writing tests, figuring out how to port the AzureDefaultCredentials functionality to Python, and tweaking the IntakePipeline approach, but I wanted to get early eyes on this to ensure it's a viable approach.

### CODE REVIEW

**Comment by JSkeets** [comment]
Has this been tested at all ? One off on your machine or otherwise ?

**Comment by bryan-skylight** [comment]
> Has this been tested at all? One off on your machine or otherwise?

No, because of as of right now it wouldn't work since the functionality to get the access token hasn't been ported to Python. This PR should have been submitted as a draft as it's not intended to be considered for merging. It was just a quick port of the main DSTP functionality to Python to help frame everyone's thoughts about what this would look like if we moved the FHIR conversion into `IntakePipeline`.

**Comment by bryan-skylight** [comment]
Just pushed a major update to this PR. The code now successfully runs locally via a BlobTrigger and is able to run the code through all of our transforms (convert to FHIR; standardize name, address, and phone number; geocode; and generate patient hash) and then save the FHIR bundle to the appropriate location. It also saves the message into an invalid messages location if the conversion fails, which is happening quite a bit due to malformed datetimes. 

We'll need to confirm that the transformations are actually happening, and we'll need to make sure we can run this in Azure, but I believe this gets us closer to the ideal scenario where all transformations, including the FHIR conversion, are done in a single pipeline.

File: src/FunctionApps/python/IntakePipeline/.gitignore
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/.gitignore:1]: Is this any different than the one in the root?

File: src/FunctionApps/python/IntakePipeline/conversion.py
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/conversion.py:131]: $0.02: Usually I like keeping things that read from config at a higher level (ie: as an arg to `convert_message_to_fhir`)

File: src/FunctionApps/python/IntakePipeline/conversion.py
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/conversion.py:151]:
```suggestion
    response = requests.post(
        url=url,
        json=data,
        headers={
            "Authorization": f"Bearer {access_token}"
        }
    )
```

File: src/FunctionApps/python/IntakePipeline/function.json
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/function.json:8]: was this specifically running on ELRs or is this just the last thing you were testing on?

File: src/FunctionApps/python/IntakePipeline/getting_started.md
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/getting_started.md:1]: I'd get rid of this

File: src/FunctionApps/python/IntakePipeline/requirements.txt
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/requirements.txt:3]: we've got reqiurements.txt in the python function app root

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (graham-skylight) [inline src/FunctionApps/python/IntakePipeline/__init__.py:34]: This is a lot to put in the major orchestrating function - can we refactor a little?

File: src/FunctionApps/python/IntakePipeline/function.json
Comment (DanPaseltiner) [inline src/FunctionApps/python/IntakePipeline/function.json:8]: Is this path correct? Don't we want he blob trigger to trigger whenever there is a new blob for any message type? What about point the trigger to just bronze/decrypted?

**Review by graham-skylight — DISMISSED** [review]
Glad to see us integrating all this! A bunch of minor points in comments and a few more:

* First things first, let's get this running in the pitest app - we've had lots of false starts with config and it'd be nice to prove that this can work there
* This drops test coverage by a good bit (and I'm pretty sure it'll fail tests once the conflicts are figured out). Is there a plan to fix it?

File: src/FunctionApps/python/IntakePipeline/getting_started.md
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/getting_started.md]: Done.

File: src/FunctionApps/python/IntakePipeline/requirements.txt
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/requirements.txt]: Removed this instance of it.

File: src/FunctionApps/python/IntakePipeline/function.json
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/function.json]: This was the one I was testing on when I pushed. I confirmed it worked with the other file types after pushing and have now pushed the proper `function.json` file, which uses `bronze/decrypted/{name}`.

File: src/FunctionApps/python/IntakePipeline/.gitignore
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/.gitignore]: Nope. Removed this instance of it.

**Comment by bryan-skylight** [comment]
> This drops test coverage by a good bit (and I'm pretty sure it'll fail tests once the conflicts are figured out). Is there a plan to fix it?

Yes, I definitely intend to write tests for this before merging. PR should be updated by EOD 4/19.

File: src/FunctionApps/python/IntakePipeline/conversion.py
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/conversion.py]: Makes sense to me. I've moved this outside of the function and now pass `fhir_url` in.

File: src/FunctionApps/python/IntakePipeline/conversion.py
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/conversion.py]: Good call! This is much cleaner and has now been implemented.

File: src/FunctionApps/python/IntakePipeline/function.json
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/function.json]: This was an accidental remnant of running tests and pushing before changing the path back. The code has been tested on all three file types and an updated `path` has been pushed to the PR.

**Review by bamader — APPROVED** [review]
Not many changes to suggest here, things look mostly good. Obviously top priority is just verifying that it actually deploys and runs.

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (spencer-skylight) [inline src/FunctionApps/python/IntakePipeline/__init__.py:72]: This method requires the bundle type to be supplied (elr, ecr, or vxu).  These values come from constants defined in the GenerateCSV Azure function and differ (but have a 1-1 relationship with) the filetype paramater.  These should probably be standardized across the solution.  In the short term, I think we'll need to pass either one of those values to this function - probably derived from the filetype parameter.  

Code that uses type:
https://github.com/CDCgov/prime-public-health-data-infrastructure/blob/6152260fe4e6f033d0e500fdcb4e587d478aa4e6/src/FunctionApps/python/IntakePipeline/__init__.py#L33

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/__init__.py]: `clean_message` expects a string to be passed in, but the data is in byte format, so when you try to decode to UTF to get a string, which would then be passed to `clean_message`, the error has already occurred. As a result, I think the comment should stay until we figure out how to remove those bytes before trying to decode it.

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (spencer-skylight) [inline src/FunctionApps/python/IntakePipeline/__init__.py:54]: Should this method be defined in conversion.py

File: src/FunctionApps/python/tests/IntakePipeline/pipeline_test.py
Comment (spencer-skylight) [inline src/FunctionApps/python/tests/IntakePipeline/pipeline_test.py:9]: This change requires a terraform update too I think.

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (spencer-skylight) [inline src/FunctionApps/python/IntakePipeline/__init__.py:104]: We should probably add a unit test for this.

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (bryan-skylight) [inline src/FunctionApps/python/IntakePipeline/__init__.py:91]: Are these values reversed from what's expected in line 110? It's unpacking the values expecting the bundle to be first, but we're returning it as the second value here. Also, I think we should change line 110 to use bundle_type to keep the variable names consistent.

Comment (bryan-skylight) [inline src/FunctionApps/python/IntakePipeline/__init__.py:91]: Are these values reversed from what's expected in line 110? It's unpacking the values expecting the bundle to be first, but we're returning it as the second value here. Also, I think we should change line 110 to use bundle_type to keep the variable names consistent.

**Review by bryan-skylight — COMMENTED** [review]
I

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (spencer-skylight) [inline src/FunctionApps/python/IntakePipeline/__init__.py:110]: This looks backwards - convert_to_fhir returns (bundle_type, bundle).

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (bryan-skylight) [inline src/FunctionApps/python/IntakePipeline/__init__.py:96]: Not returning anything here and not raising an exception results in the code continuing to run on line 110, which results an unhandled exception where too many values are trying to be unpacked. Probably easiest to just return an empty tuple.

Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/__init__.py:96]: Even correcting for this locally is still resulting in the unpacking error. I'm looking into why this is the case and will update once I know more.

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (bamader) [reply src/FunctionApps/python/IntakePipeline/__init__.py:91]: Yeah we definitely can't have a `return` here with a singular value. Based on line 107 where we iterate over a sequence of bundles, it could make sense to create a list of tuples of `(str, dict)`, i.e. `valid_processed_bundles: List[Tuple[str, dict]]. Then when we process an individual bundle, we just access the ith tuple.

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (bamader) [reply src/FunctionApps/python/IntakePipeline/__init__.py:96]: I would agree with the empty tuple return, and then we just define a check on the length of the tuple later on when processing bundles. Or, potentially, if we take the approach I outlined of building up a list of tuples, then we may not need to return anything at all here, since we'll already have a list of the fully valid and processed bundles.

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (spencer-skylight) [reply src/FunctionApps/python/IntakePipeline/__init__.py:81]: FWIW, the cred manager keeps track of an access token + expiration date.  This call will check the token (which is tracked in memory) isn't about to expire, and only get a new one if it needs to.  So the cost is relatively low to keeping it in the loop.  But Graham has also rightly suggested that it isn't necessary since the function only lives for bundle at a time the token really should never expire.  Note the downside of this is we get a new token for every bundle (or at least every file).  No strong preference if you change this - I think it will work either way.  Just adding some context on how it was designed.

File: src/FunctionApps/python/IntakePipeline/fhir.py
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/fhir.py]: Great call out. I opted to rework the code a bit and generalize the `store_bundle` function to handle both valid and invalid messages. Hopefully this addresses the concern.

**Comment by bryan-skylight** [comment]
I just pushed another big update to this PR. Here's a gist of what I changed:

* Refactored `main` and `run_pipeline` such that main is now responsible for iterating over the messages while run_pipeline will only handle one message at a time.  
  * This removes the `convert_to_fhir` function that was added in 9dab3c3, but still maintains the heart of the approach by keeping `run_pipeline` a 1:1 method while `main` handles the 1:N logic.  
  * The objective here is to keep the FHIR conversion inside of `run_pipeline` as I think that more closely represents our architecture diagram and better represents the idea of taking in the original data and processing it in its entirety from conversion to storage.   
* Removed all of the mappings for input data type, root template, and template collection from the `convert_message_to_fhir` function and placed them in a function that also includes the `if` tree logic that was being used to determine `bundle_type`.  
  * All of this logic now resides in a function that generates a dictionary for each file (not message) that is processed.  
* `store_bundle` has been renamed to `store_data` and now handles storing both valid and invalid messages.  
  * All files are now stored with a file name that refers to their parent file as well as a message index that should help us map any given valid or invalid file back to where it came from.  
  * This also removes a dependency on the `uuid` library.  
* `convert_message_to_fhir` now includes the logic for determining if the conversion failed or not, and returns either an empty dict (if the conversion failed) or the FHIR data. `run_pipeline` then simply checks if the dictionary is empty and proceeds as appropriate.

I believe these changes incorporate all of the feedback that was provided, both in the form of comments and code that was submitted.

I've tested this locally on my machine and confirmed that it works (at least insofar as it produces both valid and invalid messages and they're stored where/how they're supposed to be), and will be pushing another update tomorrow (04/20) that adds any necessary tests.

Feedback requested:
* Are there any changes that were requested that I missed?  
* Does this refactor achieve what was requested in the comments/code about refactoring the large `run_pipeline` code?  
* Is there any reason that we can't pass the access token and FHIR URL to `upload_fhir_bundle_to_server` to avoid needing to pass `cred_manager` both to `run_pipeline` and then to that function?

File: src/FunctionApps/python/tests/IntakePipeline/pipeline_test.py
Comment (graham-skylight) [inline src/FunctionApps/python/tests/IntakePipeline/pipeline_test.py:32]: this is getting to be a *lot*

File: src/FunctionApps/python/tests/IntakePipeline/pipeline_test.py
Comment (graham-skylight) [inline src/FunctionApps/python/tests/IntakePipeline/pipeline_test.py:54]: `MESSAGE_MAPPINGS` isn't being used elsewhere, right? Let's inline it (no need to build out the dict, it's a test)

**Review by graham-skylight — COMMENTED** [review]
Glad to hear we got it working locally!

This would normally be a -1, mostly because the new code (the DSTP port, not the glue code to run the pipeline) isn't covered, and I think we could stand to simplify some parts.

That said, do we just want to merge this to unblock people and fix it after the fact? We've already got 50+ comments on this.

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/__init__.py:94]: The way I thought about to solve this would be to move `get_file_type_mappings` into the `for` loop and generate the mappings for each message. This would be necessary because we need the message index so we know which message it maps back to in the parent file. Generating the mappings for each message felt like overkill, so this is a small price to pay to reduce the number of times the function gets called.

File: src/FunctionApps/python/IntakePipeline/fhir.py
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/fhir.py:20]: That's correct. Happy to change this to something that makes more sense.

File: src/FunctionApps/python/IntakePipeline/fhir.py
Comment (bryan-skylight) [reply src/FunctionApps/python/IntakePipeline/fhir.py:26]: Agreed, but do we really want two functions that do almost exactly the same thing?

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (spencer-skylight) [reply src/FunctionApps/python/IntakePipeline/__init__.py:96]: I think it's fine to pass in just the token.

File: src/FunctionApps/python/IntakePipeline/__init__.py
Comment (spencer-skylight) [reply src/FunctionApps/python/IntakePipeline/__init__.py:94]: I see both points, and probably too small a difference to worry much about.  Perhaps a middle ground would be to split the parameter into `file_type_mappings` and `message_mappings`.  The former are things that could be derived from the file, and the latter could be message-specific info (like output filename).

File: src/FunctionApps/python/IntakePipeline/fhir.py
Comment (spencer-skylight) [reply src/FunctionApps/python/IntakePipeline/fhir.py:26]: A matter of framing I suppose.  My $0.02 is having a function that can store a blob without caring whether it's a message or bundle makes more sense.

File: src/FunctionApps/python/tests/IntakePipeline/pipeline_test.py
Comment (spencer-skylight) [reply src/FunctionApps/python/tests/IntakePipeline/pipeline_test.py:32]: I think you're both right.  I feel like having a higher level orchestration component (eg. data factory) might be a better architectural direction.  This is OK for the pilot, but the more blocks we add to this, the more unwieldy it becomes to manage, and further away from intended use of a azure function.  We should foster this kind of thinking to apply post-pilot.

**Review by graham-skylight — APPROVED** [review]
:+1: tests look good

File: src/FunctionApps/python/tests/IntakePipeline/conversion_test.py
Comment (bamader) [inline src/FunctionApps/python/tests/IntakePipeline/conversion_test.py:71]: You could _potentially_ add another set of assertions for the entries in list 1 having all three group strings in them (i.e. MSH, PID, OBX each), but that's probably overkill and these tests would satisfy me as is.

File: src/FunctionApps/python/tests/IntakePipeline/pipeline_test.py
Comment (bamader) [reply src/FunctionApps/python/tests/IntakePipeline/pipeline_test.py:32]: @spencer-skylight Actually the goal for us is to provide a lot of different, plug and play independent blocks that STLTs can pick from at their leisure. It's actually a strength of our service that we're making so many things independent because it helps us deliver mission value, even if that does mean the organization gets kind of wonky.

**Review by bamader — APPROVED** [review]

**Pull request merged**
